system-contracts/contracts/BatchAggregator.sol

```diff
@@ -0,0 +1,202 @@
+        /// Check State Diffs
+        /// encoding is as follows:
+        /// header (1 byte version, 3 bytes total len of compressed, 1 byte enumeration index size)
+        /// body (`compressedStateDiffSize` bytes, 4 bytes number of state diffs, `numberOfStateDiffs` * `STATE_DIFF_ENTRY_SIZE` bytes for the uncompressed state diffs)
+        /// encoded state diffs: [20bytes address][32bytes key][32bytes derived key][8bytes enum index][32bytes initial value][32bytes final value]
+        require(
+            uint256(uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]))) ==
+                STATE_DIFF_COMPRESSION_VERSION_NUMBER,
+            "state diff compression version mismatch"
+        );
+        calldataPtr++;
+
+        uint24 compressedStateDiffSize = uint24(bytes3(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 3]));
+        calldataPtr += 3;
+
+        uint8 enumerationIndexSize = uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]));
+        calldataPtr++;
+
+        bytes calldata compressedStateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            compressedStateDiffSize];
+        calldataPtr += compressedStateDiffSize;
+
+        bytes calldata totalL2ToL1Pubdata = _totalL2ToL1PubdataAndStateDiffs[:calldataPtr];
+
+        uint32 numberOfStateDiffs = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
+        calldataPtr += 4;
+
+        bytes calldata stateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            (numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE)];
+        calldataPtr += numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE;
+
+        require(enumerationIndexSize <= MAX_ENUMERATION_INDEX_SIZE, "enumeration index size is too large");
+        uint256 numberOfInitialWrites = uint256(compressedStateDiffs.readUint16(0));
+        bytes32[] memory derivedKeyTable = new bytes32[](numberOfInitialWrites);
+
+        uint256 stateDiffPtr = 2;
+        uint256 numInitialWritesProcessed = 0;
+        // Process initial writes
+        for (uint256 i = 0; i < numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE; i += STATE_DIFF_ENTRY_SIZE) {
+            bytes calldata stateDiff = stateDiffs[i:i + STATE_DIFF_ENTRY_SIZE];
+            uint64 enumIndex = stateDiff.readUint64(84);
+            if (enumIndex != 0) {
+                // It is a repeated write, so we skip it.
+                continue;
+            }
+
+            bytes32 derivedKey = stateDiff.readBytes32(52);
+            derivedKeyTable[numInitialWritesProcessed] = derivedKey;
+            uint256 initValue = stateDiff.readUint256(92);
+            uint256 finalValue = stateDiff.readUint256(124);
+
+            numInitialWritesProcessed++;
+            require(derivedKey == compressedStateDiffs.readBytes32(stateDiffPtr), "iw: initial key mismatch");
+            if (keyStatus[chainId][derivedKey] == false) {
```

Comment:
I understand it is draft, just in case better to rename to `isKeyTouched`

Reply:
done